lms/lib/courseware_search/lms_filter_generator.py

```diff
@@ -8,37 +8,88 @@
 from opaque_keys.edx.keys import CourseKey
 from opaque_keys.edx.locations import SlashSeparatedCourseKey
+from xmodule.modulestore.django import modulestore
 
 from search.filter_generator import SearchFilterGenerator
-from openedx.core.djangoapps.course_groups.partition_scheme import get_cohorted_user_partition
+from openedx.core.djangoapps.user_api.partition_schemes import RandomUserPartitionScheme
+from openedx.core.djangoapps.course_groups.partition_scheme import CohortPartitionScheme
 from courseware.access import get_user_role
 
 
+INCLUDE_SCHEMES = [CohortPartitionScheme, RandomUserPartitionScheme, ]
+SCHEME_SUPPORTS_ASSIGNMENT = [RandomUserPartitionScheme, ]
+
+
 class LmsSearchFilterGenerator(SearchFilterGenerator):
     """ SearchFilterGenerator for LMS Search """
 
+    _user_enrollments = {}
+
+    def _enrollments_for_user(self, user):
+        """ Return the specified user's course enrollments """
+        if user not in self._user_enrollments:
+            self._user_enrollments[user] = CourseEnrollment.enrollments_for_user(user)
+        return self._user_enrollments[user]
+
     def filter_dictionary(self, **kwargs):
-        """ base implementation which filters via start_date """
-        filter_dictionary = super(LmsSearchFilterGenerator, self).filter_dictionary(**kwargs)
-        if 'user' in kwargs and 'course_id' in kwargs and kwargs['course_id']:
-            user = kwargs['user']
-            try:
-                course_key = CourseKey.from_string(kwargs['course_id'])
-            except InvalidKeyError:
-                course_key = SlashSeparatedCourseKey.from_deprecated_string(kwargs['course_id'])
-
-            # Staff user looking at course as staff user
-            if get_user_role(user, course_key) == 'staff':
-                return filter_dictionary
-
-            cohorted_user_partition = get_cohorted_user_partition(course_key)
-            if cohorted_user_partition:
-                partition_group = cohorted_user_partition.scheme.get_group_for_user(
+        """ LMS implementation, adds filtering by user partition, course id and user """
+
+        def get_group_for_user_partition(user_partition, course_key, user):
+            """ Returns the specified user's group for user partition """
+            if user_partition.scheme in SCHEME_SUPPORTS_ASSIGNMENT:
+                return user_partition.scheme.get_group_for_user(
+                    course_key,
+                    user,
+                    user_partition,
+                    assign=False,
+                )
+            else:
+                return user_partition.scheme.get_group_for_user(
                     course_key,
                     user,
-                    cohorted_user_partition,
+                    user_partition,
                 )
-                filter_dictionary['content_groups'] = unicode(partition_group.id) if partition_group else None
+
+        def get_groups_for_user(course, user):
```

Comment:
I now see that this returns group ids, not groups, so I'd rename it to ``get_group_ids_for_user``.